### DISCUSSION TITLE
Pause pipewire output stream when no clients connected

### DISCUSSION THREAD

**Discussion by Nemo157** [mikebrady/shairport-sync#1889] [Ideas]
I have my system configured to sleep when idle, but that means if I'm just playing audio through it via shairport-sync it will go to sleep and disconnect the client. There is this utility that watches pipewire status and inhibits idle state while there is any audio playing: https://github.com/rafaelrc7/wayland-pipewire-idle-inhibit. Unfortunately shairport-sync's output stream stays active all the time even without any clients so that blocks going to sleep at all. I'm not sure how it's done through pipewire's api, but with mpd when I pause the music the stream stays connected but goes into a `paused` state which then allows the idle state:

```
        55. Music Player Daemon
             60. output_FL       > WH-1000XM3:playback_FL       [paused]
            111. output_FR       > WH-1000XM3:playback_FR       [paused]
        93. Shairport Sync
             91. output_FL       > WH-1000XM3:playback_FL       [active]
            102. output_FR       > WH-1000XM3:playback_FR       [active]
```

**Comment by mikebrady** [comment]
Thanks for the report and the information. It’s something to figure out alright.

There may be another problem, however. Assuming we can figure out the PipeWire pause stuff, Shairport Sync has to stay active on the network to be able to receive new connections, etc. Can that be done while your system is in sleep mode?

**Reply by Nemo157** [reply]
No, I'm not too concerned with having to wake it up if I want to start streaming to it, it's just the going to sleep after I've connected and am listening to something that is annoying me right now.

**Reply by mikebrady** [reply]
Thanks. I'll have to experiment a bit with the PipeWire API.

**Reply by microfx** [reply]
as I wrote some time back, I need to find a way as well https://github.com/mikebrady/shairport-sync/discussions/1885

(I need to orchestrate / watch two / three possible situations where I don't want something similar — just switch a powersocket instead of sleep - but kind of similar enough?) 

Maybe we can solve both of our needs with one smart idea (I'm not yet using pipewire though). Is there a way to read from dbus?

**Reply by mikebrady** [reply]
@microfx, I really don't think this is related to your requirements -- what's happening here is that the PipeWire backend never stops sending audio, even when it is quiet.

However, you should be able to read Shairport Sync's `active` status from the D-Bus interface. If you have a Linux with a GUI, you can use D-Feet to examine what the interface provides.

There's a C-code sample program that watches for changes on Shairport Sync's D-Bus interface and prints something out. I imagine -- but don't know and have not tried -- that there are commands for monitoring D-Bus you could use in a script.

Add `--with-dbus-test-client` to the build of Shairport Sync (I know -- it's a dirty rotten hack) and the client will be built along with Shairport Sync itself. Some of it, like setting the volume control, may not work, but it's worth playing with it.

**Comment by mikebrady** [comment]
@Nemo157, can you say how you're generating the listing above please. This one:
```
        55. Music Player Daemon
             60. output_FL       > WH-1000XM3:playback_FL       [paused]
            111. output_FR       > WH-1000XM3:playback_FR       [paused]
        93. Shairport Sync
             91. output_FL       > WH-1000XM3:playback_FL       [active]
            102. output_FR       > WH-1000XM3:playback_FR       [active]
```

**Reply by Nemo157** [reply]
`wpctl status`

**Reply by mikebrady** [reply]
Yikes, we learn something new every day!

**Comment by mikebrady** [comment]
OK, so an update has been pushed into the `development` branch which implements this idea in a rather basic way. If you got a chance to test it and report back, it would be appreciated.

**Reply by Nemo157** [reply]
Yes, that works, I see the status update in `wpctl status` and `wayland-pipewire-idle-inhibit` responds correctly

```
PWGraph::update] Updated Link (102) from LinkData { input_port: Some(53), output_port: Some(177), active: Some(false) }
PWGraph::update] Updated Link (102) to LinkData { input_port: Some(53), output_port: Some(177), active: Some(true) }
...
InhibitIdleState::set_is_idle_inhibited] Started Timer to inhibit idling
WaylandIdleInhibitor::set_inhibit_idle] Idle Inhibitor was ENABLED
...
PWGraph::update] Updated Link (102) from LinkData { input_port: Some(53), output_port: Some(177), active: Some(true) }
PWGraph::update] Updated Link (102) to LinkData { input_port: Some(53), output_port: Some(177), active: Some(false) }
...
WaylandIdleInhibitor::set_inhibit_idle] Idle Inhibitor was DISABLED
```

The only issue I see with it is it starts in `active` state before any client connects.

**Reply by mikebrady** [reply]
Thanks. I’ll take a look at that starting behaviour.

**Reply by mikebrady** [reply]
That starting behaviour should (?) be fixed now. It should start in the inactive state.

**Reply by Nemo157** [reply]
Yep, it looks like it works perfectly now.

**Reply by mikebrady** [reply]
Thanks for the update.
